src/girouette/tw/border.cljc

```diff
@@ -1,10 +1,204 @@
+    :rules    "
+    divide-color = <'divide-'> color
+    "
+    :pipeline (assoc default-pipeline
+                :class-name [(fn [rule props]
+                               [(gs/> (dot (:class-name props)) (gs/+ :* :*)) rule])])
+    :garden   (fn [{[[_ color]] :component-data}]
+                (let [color (read-color color)]
+                  (if (string? color)
+                    {:border-color color}
+                    (let [[r g b a] color]
+                      (if (some? a)
+                        {:border-color (color->css color)}
+                        {:--gi-divide-opacity 1
+                         :border-color        (color->css [r g b "var(--gi-divide-opacity)"])})))))}
+
+   {:id     :divide-opacity
+    :rules  "
+    divide-opacity = <'divide-opacity-'> number
+    "
+    :garden (fn [{[[_ value]] :component-data}]
+              {:--gi-divide-opacity (value->css (/ (read-number value) 100.0))})}
+
+   {:id       :divide-style
+    :rules    "
+    divide-style = <'divide-'> divide-style-name
+    divide-style-name = 'solid' | 'dashed' | 'dotted' | 'double' | 'none'
+    "
+    :pipeline (assoc default-pipeline
+                :class-name [(fn [rule props]
+                               [(gs/> (dot (:class-name props)) (gs/+ :* :*)) rule])])
+    :garden   (fn [{[[_ border-style]] :component-data}]
+                {:border-style border-style})}
+
+   {:id     :ring-width
+    :rules  "
+    ring-width = <'ring'> (<'-'> ring-width-size)?
+    ring-width-size = 'inset' | number | length | length-unit
+    "
+    :garden (fn [{[[_ size]] :component-data}]
+              (if (= size "inset")
+                {:--gi-ring-inset "inset"}
+                (let [size (if (nil? size)
+                             "3px"
+                             (value-unit->css nil size {:number-unit "px"}))]
+                  {:box-shadow (str "var(--gi-ring-inset) 0 0 0 calc("
+                                    size
+                                    " + var(--gi-ring-offset-width))"
+                                    " var(--gi-ring-color)")})))}
+
+   {:id     :ring-color
+    :rules  "
+    ring-color = <'ring-'> color
+    "
+    :garden (fn [{[[_ color]] :component-data}]
+              (let [color (read-color color)]
+                (if (string? color)
+                  {:--gi-ring-color color}
+                  (let [[r g b a] color]
+                    (if (some? a)
+                      {:--gi-ring-color (color->css color)}
+                      {:--gi-ring-color (color->css [r g b "var(--gi-ring-opacity)"])})))))}
+   {:id     :ring-opacity
+    :rules  "
+    ring-opacity = <'ring-opacity-'> number
+    "
+    :garden (fn [{[[_ value]] :component-data}]
+              {:--gi-ring-opacity (value->css (/ (read-number value) 100.0))})}
+
+   {:id     :ring-offset-width
+    :rules  "
+    ring-offset-width = <'ring-offset-'> ring-offset-size
+    ring-offset-size = number | length | length-unit
+    "
```

Comment:
1. `ring-offset-size` should be named `ring-offset-width-size`
2. I think it is simpler to not have a separate rule, I suggest using a 1 line grammar instead.

```suggestion
    :rules  "
    ring-offset-width = <'ring-offset-'> (number | length | length-unit)
    "
```

Reply:
Not a problem to rename it if you'd like, but I think here we can stay consistent with the section title from the docs (i.e. [`ring-offset-width`](https://tailwindcss.com/docs/ring-offset-width)).

Reply:
What I meant is that the prefix `ring-offset` you used in `ring-offset-size`was not the name of the rule `ring-offset-width`.

Reply:
Gotcha; you're absolutely right. I was just bothered by the redundancy of "width" + "size"… but that's no longer an issue with the one-liner.